craft_application/git/_git_repo.py

```diff
@@ -0,0 +1,310 @@
+        """
+        self.path = path
+
+        if not path.is_dir():
+            raise FileNotFoundError(
+                f"Could not initialize a git repository because {str(path)!r} does not "
+                "exist or is not a directory."
+            )
+
+        if not is_repo(path):
+            self._init_repo()
+
+        self._repo = pygit2.Repository(path.as_posix())
+
+    def add_all(self) -> None:
+        """Add all changes from the working tree to the index.
+
+        :raises GitError: if the changes could not be added
+        """
+        logger.debug("Adding all changes.")
+
+        try:
+            self._repo.index.add_all()  # pyright: ignore[reportUnknownMemberType]
+            self._repo.index.write()  # pyright: ignore[reportUnknownMemberType]
+        except pygit2.GitError as error:
+            raise GitError(
+                f"Could not add changes for the git repository in {str(self.path)!r}."
+            ) from error
+
+    def commit(self, message: str = "auto commit") -> str:
+        """Commit changes to the repo.
+
+        :param message: the commit message
+
+        :returns: object ID of the commit as str
+
+        :raises GitError: if the commit could not be created
+        """
+        logger.debug("Committing changes.")
+
+        try:
+            tree = (
+                self._repo.index.write_tree()  # pyright: ignore[reportUnknownMemberType]
+            )
+        except pygit2.GitError as error:
+            raise GitError(
+                f"Could not create a tree for the git repository in {str(self.path)!r}."
+            ) from error
+
+        author = pygit2.Signature("auto commit", "auto commit")
+
+        # a target is not needed for an unborn head (no existing commits in branch)
+        target = [] if self._repo.head_is_unborn else [self._repo.head.target]
+
+        try:
+            return str(
+                self._repo.create_commit("HEAD", author, author, message, tree, target)
+            )
+        except pygit2.GitError as error:
+            raise GitError(
+                "Could not create a commit for the git repository "
+                f"in {str(self.path)!r}."
+            ) from error
+
+    def is_clean(self) -> bool:
+        """Check if the repo is clean.
+
+        :returns: True if the repo is clean.
+
+        :raises GitError: if git fails while checking if the repo is clean
+        """
+        try:
+            # for a clean repo, `status()` will return an empty dict
+            return not bool(self._repo.status())
+        except pygit2.GitError as error:
+            raise GitError(
+                f"Could not check if the git repository in {str(self.path)!r} is clean."
+            ) from error
+
+    def _init_repo(self) -> None:
+        """Initialize a git repo.
+
+        :raises GitError: if the repo cannot be initialized
+        """
+        logger.debug("Initializing git repository in %r", str(self.path))
```

Comment:
It could be, but we intentionally do it in a hidden place. It's debug because this is an implementation detail of how remote build works rather than something to be expected